### PR TITLE
Unroll

### PR DESCRIPTION
**Pull request by williangaspar**
Fixes: https://github.com/iovisor/bpftrace/issues/9

### CODE REVIEW

File: src/ast/codegen_llvm.cpp [hunk 1]
```diff
@@ -962,6 +962,36 @@ void CodegenLLVM::visit(If &if_block)
...
+void CodegenLLVM::visit(Unroll &unroll)
+{
+  Function *parent = b_.GetInsertBlock()->getParent();
+  BasicBlock *loop = BasicBlock::Create(module_->getContext(), "loop", parent);
+  BasicBlock *done = BasicBlock::Create(module_->getContext(), "done", parent);
+
+  AllocaInst *val = b_.CreateAllocaBPF(SizedType(Type::integer, 8), "loop_count");
+  b_.CreateStore(b_.getInt64(unroll.var), val);
+
+  expr_ = val;
```
Comment (ajor) [inline src/ast/codegen_llvm.cpp]: expr_ is just intended for passing expressions back up the tree. You could just use val in the comparison below

File: src/ast/codegen_llvm.cpp [hunk 2]
```diff
@@ -962,6 +962,36 @@ void CodegenLLVM::visit(If &if_block)
...
+  }
+
+  Value *var = b_.CreateLoad(val);
+  expr_ = b_.CreateSub(var, b_.getInt64(1), "subtmp");
+  b_.CreateStore(expr_, val);
+
+  b_.CreateCondBr(b_.CreateICmpNE(expr_, b_.getInt64(0), "true_cond"), loop, done);
+
+  b_.SetInsertPoint(done);
+  expr_ = nullptr;
```
Comment (ajor) [inline src/ast/codegen_llvm.cpp]: Could you stick in a lifetime end for val just before the end here

**Comment by brendangregg** [comment]
I get the same errors as @ajor; @williangaspar: what llvm/clang version are you using?

**Comment by williangaspar** [comment]
> I get the same errors as @ajor; @williangaspar: what llvm/clang version are you using?

I'm using llvm 5 on ubuntu 18.4.

**Comment by brendangregg** [comment]
If I change this:

`  b_.CreateCondBr(b_.CreateICmpNE(val, b_.getInt64(0), "true_cond"), loop, done);`

to this:

`  b_.CreateCondBr(b_.CreateICmpNE(b_.getInt64(unroll.var), b_.getInt64(0), "true_cond"), loop, done);`

It works:

```
# ./src/bpftrace -e 'kprobe:do_nanosleep { $i = 0; unroll(5) { printf("i: %d\n", $i); $i = $i + 1; } }'
Attaching 1 probe...
i: 0
i: 1
i: 2
i: 3
i: 4
i: 0
i: 1
i: 2
i: 3
i: 4
^C
```

I believe it was crashing during compile time:

```
# ./src/bpftrace -d -e 'kprobe:do_nanosleep { $i = 0; unroll(5) { printf("i: %d\n", $i); $i = $i + 1; } }'
Program
 kprobe:do_nanosleep
  =
   variable: $i
   int: 0
  unroll 5
   call: printf
    string: i: %d\n
    variable: $i
   =
    variable: $i
    +
     variable: $i
     int: 1

bpftrace: /usr/lib/llvm-5.0/include/llvm/IR/Instructions.h:1113: void llvm::ICmpInst::AssertOK(): Assertion `getOperand(0)->getType() == getOperand(1)->getType() && "Both operands to ICmp instruction are not of the same type!"' failed.
Aborted (core dumped)
```

**Comment by brendangregg** [comment]
Thanks Willian!

**Pull request merged**